### PR TITLE
Feat: Client Reports

### LINKED ISSUE #735: Add 'Client Reports' to the Dart SDK
Sentry's protocol now offers a way to report counts and reason to why data gets dropped.

Spec: https://develop.sentry.dev/sdk/client-reports/

This was implemented in other places that can be used as a reference:
* JavaScript: https://github.com/getsentry/sentry-javascript/pull/3955
* Python: https://github.com/getsentry/sentry-python/pull/1181
* Relay: https://github.com/getsentry/relay/pull/1074

This will also be implemented in the following SDKs so collaborating with folks adding support there is ideal
* .NET: https://github.com/getsentry/sentry-dotnet/issues/1455
* Java: https://github.com/getsentry/sentry-java/issues/1894
* Apple: https://github.com/getsentry/sentry-cocoa/issues/1655
* React Native: https://github.com/getsentry/sentry-react-native/issues/2034

### LINKED ISSUE #812: `SentryEnvelopeItem.fromMethods` should always use `_CachedItem` to do it async in case users don't await for the Future to be completed
https://github.com/getsentry/sentry-dart/pull/810/files#r837490293

### PR DESCRIPTION
**Pull request by denrase**
## :scroll: Description

Feature branch collecting all sub-branches for the client reports implementation.

## :bulb: Motivation and Context

- [x] Feat: Record & Send Client Reports #813
- [x] Feat: Client Report Envelope Item #810
- [x] Feat: Client Report Recorder #809
- [x] Client Reports: Record Network Errors #830 
- [x] Client Reports: Bump iOS SDK & Pass Options #836
- [x] Client Reports: Bump and Pass Options (Android) #846

Closes #812

## :green_heart: How did you test it?

Unit tests.

## :pencil: Checklist

- [x] I reviewed submitted code
- [x] I added tests to verify changes
- [ ] I updated the docs if needed
- [x] All tests passing
- [x] No breaking changes

## :crystal_ball: Next steps

- [x] Set `sendClientReports` on native Android options when [Java Client Reports](https://github.com/getsentry/sentry-java/pull/1982) are merged
- [ ] Test End-To-End on iOS, Android and Web

Closes #735

### CODE REVIEW

File: dart/lib/src/sentry_envelope_item.dart [hunk 1]
```diff
@@ -12,21 +13,12 @@ class SentryEnvelopeItem {
...
-        transaction.toJson(),
-        toEncodable: jsonSerializationFallback,
-      );
-      return utf8.encode(jsonEncoded);
-    });
-
-    final getLength = () async {
-      return (await cachedItem.getData()).length;
-    };
+    final cachedItem =
```
Comment (marandaneto) [inline dart/lib/src/sentry_envelope_item.dart:16]: If we are using always CachedItem, we can close https://github.com/getsentry/sentry-dart/issues/812 ?

File: dart/lib/src/sentry_client.dart [hunk 2]
```diff
@@ -34,10 +37,13 @@ class SentryClient {
 
   /// Instantiates a client using [SentryOptions]
   factory SentryClient(SentryOptions options) {
+    if (options.sendClientReports) {
+      options.recorder = ClientReportRecorder(options.clock);
+    }
     if (options.transport is NoOpTransport) {
-      options.transport = HttpTransport(options, RateLimiter(options.clock));
+      final rateLimiter = RateLimiter(options.clock, options.recorder);
```
Comment (marandaneto) [inline dart/lib/src/sentry_client.dart]: Maybe `RateLimiter` takes `options` only and it uses `clock` and `recorder` from it

File: dart/lib/src/transport/http_transport.dart [hunk 3]
```diff
@@ -28,10 +33,10 @@ class HttpTransport implements Transport {
       options.httpClient = Client();
     }
 
-    return HttpTransport._(options, rateLimiter);
+    return HttpTransport._(options, rateLimiter, options.recorder);
```
Comment (marandaneto) [inline dart/lib/src/transport/http_transport.dart]: If we already pass the `options`, we dont need to pass the `recorder` as well.

**Review by marandaneto — APPROVED** [review]

**Comment by marandaneto** [comment]
It needs testing with the iOS and Android bridges, but codewise looks good.

**Review by brustolin — APPROVED** [review]
LGTM

File: CHANGELOG.md [hunk 4]
```diff
@@ -1,5 +1,8 @@
 # Changelog
 
+## Unreleased
+
+* Feat: Client Reports (#829)
 ## 6.5.1
```
Comment (marandaneto) [inline CHANGELOG.md]:
```suggestion
* Feat: Client Reports (#829)

## 6.5.1
```

File: CHANGELOG.md [hunk 5]
```diff
@@ -1,5 +1,9 @@
 # Changelog
 
+## Unreleased
+
+* Feat: Client Reports (#829)
```
Comment (marandaneto) [inline CHANGELOG.md:5]: Lets add a note about the breaking change as well, See on Android https://github.com/getsentry/sentry-java/pull/2006
Copy and paste is fine

**Comment by denrase** [comment]
@marandaneto Client reports are being send with the `HttpTransport` and we get a 200 response from the backend.

Bildschirmfoto 2022-04-27 um 13 59 26

**Pull request merged**